mm2src/coins/z_coin/z_tx_history.rs

```diff
@@ -0,0 +1,217 @@
+use crate::z_coin::{ZCoin, ZTxHistoryError};
+use common::PagingOptionsEnum;
+use mm2_err_handle::prelude::MmError;
+
+cfg_wasm32!(
+    use crate::z_coin::storage::wasm::tables::{WalletDbBlocksTable, WalletDbReceivedNotesTable, WalletDbTransactionsTable};
+    use crate::MarketCoinOps;
+    use mm2_number::BigInt;
+    use num_traits::ToPrimitive;
+);
+
+cfg_native!(
+    use crate::z_coin::BLOCKS_TABLE;
+    use db_common::sqlite::sql_builder::{name, SqlBuilder, SqlName};
+    use db_common::sqlite::rusqlite::Error as SqliteError;
+    use db_common::sqlite::rusqlite::Row;
+    use db_common::sqlite::offset_by_id;
+    use common::async_blocking;
+);
+
+#[cfg(not(target_arch = "wasm32"))]
+const TRANSACTIONS_TABLE: &str = "transactions";
+
+pub(crate) struct ZCoinTxHistoryItem {
+    pub(crate) tx_hash: Vec<u8>,
+    pub(crate) internal_id: i64,
+    pub(crate) height: i64,
+    pub(crate) timestamp: i64,
+    pub(crate) received_amount: i64,
+    pub(crate) spent_amount: i64,
+}
+
+pub(crate) struct ZTxHistoryRes {
+    pub(crate) total_tx_count: u32,
+    pub(crate) transactions: Vec<ZCoinTxHistoryItem>,
+    pub(crate) skipped: usize,
+}
+
+/// Fetch transaction history from the database.
+#[cfg(target_arch = "wasm32")]
+pub(crate) async fn fetch_tx_history_from_db(
+    z: &ZCoin,
+    limit: usize,
+    paging_options: PagingOptionsEnum<i64>,
+) -> Result<ZTxHistoryRes, MmError<ZTxHistoryError>> {
+    let wallet_db = z.z_fields.light_wallet_db.clone();
+    let wallet_db = wallet_db.db.lock_db().await.unwrap();
+    let db_transaction = wallet_db.get_inner().transaction().await?;
+    let tx_table = db_transaction.table::<WalletDbTransactionsTable>().await?;
+    let total_tx_count = tx_table.count_all().await? as u32;
+    let offset = match paging_options {
+        PagingOptionsEnum::PageNumber(page_number) => ((page_number.get() - 1) * limit) as i64,
+        PagingOptionsEnum::FromId(tx_id) => tx_id,
+    };
+
+    // Fetch transactions
+    let txs = tx_table
+        .cursor_builder()
+        .only("ticker", z.ticker())?
+        .offset(offset as u32)
+        .limit(limit)
+        .reverse()
+        .open_cursor("ticker")
+        .await?
+        .collect()
+        .await?;
+
+    // Fetch received notes
+    let rn_table = db_transaction.table::<WalletDbReceivedNotesTable>().await?;
+    let received_notes = rn_table
+        .cursor_builder()
+        .only("ticker", z.ticker())?
+        .open_cursor("ticker")
+        .await?
+        .collect()
+        .await?;
+
+    // Fetch blocks
+    let blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+    let blocks = blocks_table
+        .cursor_builder()
+        .only("ticker", z.ticker())?
+        .open_cursor("ticker")
+        .await?
+        .collect()
+        .await?;
```

Comment:
I guess blocks are the opposite where we should use reverse
https://github.com/KomodoPlatform/komodo-defi-framework/blob/601324a0bc9c25ba0b73f136b368ba0d70071c1c/mm2src/coins/z_coin/z_tx_history.rs#L198
You should probably add tests for both wasm and non wasm `fetch_tx_history_from_db` using `cross_test` to make sure they produce the same results.

Reply:
this require modifying zcash lib to export in memory sql db for test